DMCompiler/DM/Visitors/DMProcBuilder.cs

```diff
@@ -406,8 +407,25 @@ public void ProcessStatementSet(DMASTProcStatementSet statementSet)
         }
 
         public void ProcessStatementForList(DMASTProcStatementForList statementForList) {
-            DMExpression.Emit(_dmObject, _proc, statementForList.List);
-            _proc.CreateListEnumerator();
+
+            DMASTProcStatementVarDeclaration varDeclaration = statementForList.Initializer as DMASTProcStatementVarDeclaration;
+
+            if (varDeclaration?.Type != null)
+            {
+                if (statementForList.List is null) // This shouldn't happen
+                {
+                    DMCompiler.Error(new CompilerError(varDeclaration.Location, "Attempted to create a list enumerator with a null list"));
+                    return;
+                }
+                DMExpression.Emit(_dmObject, _proc, statementForList.List);
+                _proc.CreateListEnumerator();
+            }
+            else
+            {
+                DMExpression.Emit(_dmObject, _proc, statementForList.List);
+                _proc.CreateListEnumerator();
+            }
+
```

Comment:
Code is being repeated in both the `if` and `else` blocks
```suggestion
            DMASTProcStatementVarDeclaration varDeclaration = statementForList.Initializer as DMASTProcStatementVarDeclaration;

            DMExpression.Emit(_dmObject, _proc, statementForList.List);
            _proc.CreateListEnumerator();
```